### PR TITLE
feat: add credentialProvider option when creating clients

### LINKED ISSUE #821: Handle enabling/changing redis password without application restart
It seems to be not possible to enable or change redis password without application restart.

Even in case when I want to enable password and I supply password to the driver (it logs warning that the password is not required), when the password is enabled in redis, the driver returns error "not authenticated"; it doesn't try to authenticate using password that was passed in config.

For changing password there is no option. It would not be required if redis supported user accounts - you'd be able to simply create another account, switch application to use it and then remove the old account. Redis doesn't and apparently will not have accounts.

Of course the issue can be resolved by creating a wrapper around redis driver that would create a new client when authentication fails, but it seems as either too much of a change in applications or too much of an effort to make the wrapper that behaves exactly as the driver...

Maybe there is a simpler solution to this problem, like subscribing to some event, authenticating and replaying the last command when authentication fails? Please let me know if there is.

Otherwise it would be nice to have these two features inside driver:

1. Handle enabling password:
   1) If the password is passed
   2) and it was not required by redis
   3) and at some later time some command was rejected as not authenticated
   4) driver will try authenticating again using the passed password
   5) if authentication is successful, it will resend command
   6) if authentication failed it would return error

2. Support feature of password change using new_password option (or something like this). In this case when authentication fails, the driver will try authenticating using the new password and if it was successful the driver will use it afterwards in case of re-connects.

If you are OK with these features I could implement.

Please let me know.
Thank you.

### PR DESCRIPTION
**Pull request by dhensby**
### Description

In some instances, credentials for the redis client will be short-lived and need to be fetched on-demand when connecting to redis. This is the case when connecting in AWS using IAM authentication or Entra ID in Azure.

This feature allows for a credentialProvider to be provided which is a callable function returning a Promise that resolves to a username/password object.

Closes #821

A potential enhancement to this feature could be to also return a TTL with the credentials and then automatically attempt to re-auth the client before the credentials expire; however I've assumed that logic will be the responsibility of the consumer rather than the library. Worst case scenario is that the socket disconnects due to authentication error and then re-authenticates on a future reconnection retry attempt.

Something else that is missing is an explicit decision on how errors from the credentialProvider should be surfaced.

---

### Checklist

- [ ] Does `npm test` pass with this change (including linting)?
- [x] Is the new or changed code fully tested?
- [x] Is a documentation update included (if this change modifies existing APIs, or introduces new ones)?

### CODE REVIEW

**Review by zachary-blackbird — APPROVED** [review]

File: packages/client/lib/client/index.spec.ts [hunk 1]
```diff
@@ -103,6 +104,21 @@ describe('Client', () => {
...
+
+    testUtils.testWithClient('should accept a credentialSupplier', async client => {
+      assert.equal(
+          await client.ping(),
+          'PONG'
+      );
+    }, {
+      ...GLOBAL.SERVERS.PASSWORD,
+      clientOptions: {
+        // simulate a slight pause to fetch the credentials
```
Comment (leibale) [inline packages/client/lib/client/index.spec.ts:116]: Does "simulating a pause" make the test better in any way? can't we just `Promise.resolve` directly?

**Review by leibale — COMMENTED** [review]
Overall looks good, we might have a different name for it

cc @bobymicroby

File: packages/client/lib/client/index.spec.ts (hunk 1 above)
Comment (dhensby) [reply packages/client/lib/client/index.spec.ts:116]: I found that the order of the resolution of promises could cause a problem if it was just an instantly resolving promise, that edge case wouldn't be covered.

**Comment by bobymicroby** [comment]
Hi @dhensby, thank you very much for the PR! We (Redis) are currently working on enabling not only async credential providers, but credential providers that also support rolling credentials/credential invalidation. We need them to enable EntraID (former Azure Active Directory) and more 3rd party identity providers in the future. We have a [PR](https://github.com/redis/node-redis/pull/2877) in the works that already covers async credential providers (check the first [commit](https://github.com/redis/node-redis/pull/2877/commits/a0c324b9a0e9e11f82256748b93a0f97816f482b)). It is similar to what you already have. We will be very happy to get your feedback.

**Comment by dhensby** [comment]
@bobymicroby great, that's exactly what I'm after (EntraID support), so glad to see that's in the works.

I'll close this as superseded by #2877

**Pull request closed without merging**